File: development/ui/src/app/map/planner/plan/plan-route.ts
import {List} from "immutable";
import Coordinate from "ol/coordinate";
import {PlanNode} from "./plan-node";
import {PlanSegment} from "./plan-segment";

export class PlanRoute {

  constructor(readonly source: PlanNode,
              readonly sink: PlanNode,
              readonly meters: number,
              readonly segments: List<PlanSegment>,
              readonly streets: List<string>) {
  }

  coordinates(): List<Coordinate> {
    return List([this.source.coordinate]).concat(this.segments.flatMap(segment => segment.fragments.map(fragment => fragment.coordinate)));
  }

}

File: server/src/main/scala/kpn/api/common/planner/PlanRoute.scala
package kpn.api.common.planner

case class PlanRoute(
  sourceNode: PlanNode,
  sinkNode: PlanNode,
  meters: Long,
  segments: Seq[PlanSegment],
  streets: Seq[String]
) {

  def reverse: PlanRoute = {

    if (segments.isEmpty) {
      PlanRoute(
        sinkNode,
        sourceNode,
        meters,
        Seq.empty,
        streets
      )
    }
    else {
      val fragmentCoordinates = (
        Seq(PlanFragmentCoordinate(sourceNode.coordinate, sourceNode.latLon))
          ++ segments.flatMap(_.fragments).map(f => PlanFragmentCoordinate(f.coordinate, f.latLon))
        ).reverse.drop(1)

      val fragmentCoordinateIterator = fragmentCoordinates.iterator

      val reversedSegments = segments.reverse.map { segment =>
        val reversedFragments = segment.fragments.reverse.map { fragment =>
          val fragmentCoordinate = fragmentCoordinateIterator.next()
          fragment.copy(
            latLon = fragmentCoordinate.latLon,
            coordinate = fragmentCoordinate.coordinate
          )
        }
        segment.copy(fragments = reversedFragments)
      }

      PlanRoute(
        sinkNode,
        sourceNode,
        meters,
        reversedSegments,
        streets
      )
    }
  }
}
